src/tools/clippy/clippy_lints/src/significant_drop_in_scrutinee.rs

```diff
@@ -0,0 +1,369 @@
+    expr: &'tcx Expr<'tcx>,
+) -> Option<Vec<(Span, LintSuggestion)>> {
+    let mut helper = SigDropHelper::new(cx);
+    match expr.kind {
+        ExprKind::Match(match_expr, _, _) => helper.find_sig_drop(match_expr),
+        _ => None,
+    }
+}
+
+struct SigDropHelper<'a, 'tcx> {
+    cx: &'a LateContext<'tcx>,
+    is_chain_end: bool,
+    seen_types: FxHashSet<Ty<'tcx>>,
+    has_significant_drop: bool,
+    sig_drop_spans: Option<Vec<Span>>,
+    special_handling_for_binary_op: bool,
+}
+
+#[derive(Debug)]
+enum LintSuggestion {
+    MoveOnly,
+    MoveAndDerefToCopy,
+    MoveAndClone,
+}
+
+impl<'a, 'tcx> SigDropHelper<'a, 'tcx> {
+    fn new(cx: &'a LateContext<'tcx>) -> SigDropHelper<'a, 'tcx> {
+        SigDropHelper {
+            cx,
+            is_chain_end: true,
+            seen_types: FxHashSet::default(),
+            has_significant_drop: false,
+            sig_drop_spans: None,
+            special_handling_for_binary_op: false,
+        }
+    }
+
+    fn find_sig_drop(&mut self, match_expr: &'tcx Expr<'_>) -> Option<Vec<(Span, LintSuggestion)>> {
+        self.visit_expr(match_expr);
+        if self.has_significant_drop {
+            if let Some(spans) = &self.sig_drop_spans {
+                Some(
+                    spans
+                        .into_iter()
+                        .map(|s| (s.clone(), LintSuggestion::MoveOnly))
+                        .collect::<Vec<(Span, LintSuggestion)>>(),
+                )
+            } else {
+                // If sig drop spans is empty, it means that we didn't find a type that was
+                // trivially copyable as we moved up the chain after finding a significant drop, so
+                // move the entire scrutinee.
+                let ty = self.get_type(match_expr);
+                if ty.is_ref() {
+                    let TypeAndMut { ty, .. } = ty
+                        .builtin_deref(true)
+                        .expect("Checked that the type was ref, so builtin_deref will return Some TypeAndMut");
+                    if ty.is_trivially_pure_clone_copy() {
+                        Some(vec![(match_expr.span, LintSuggestion::MoveAndDerefToCopy)])
+                    } else {
+                        Some(vec![(match_expr.span, LintSuggestion::MoveAndClone)])
+                    }
+                } else {
+                    Some(vec![(match_expr.span, LintSuggestion::MoveOnly)])
+                }
+            }
+        } else {
+            None
+        }
+    }
+
+    fn get_type(&self, ex: &'tcx Expr<'_>) -> Ty<'tcx> {
+        self.cx.typeck_results().expr_ty(ex)
+    }
+
+    fn is_trivially_copyable(&self, ex: &'tcx Expr<'_>) -> bool {
+        let ty = self.get_type(ex);
+        ty.is_trivially_pure_clone_copy()
+    }
+
+    fn has_seen_type(&mut self, ty: Ty<'tcx>) -> bool {
+        !self.seen_types.insert(ty)
+    }
+
+    fn has_sig_drop_attr(&mut self, cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> bool {
```

Comment:
I think it could -- I was trying to use the rustc_hir Visitor, but I could not figure out how to get the hir Ty from Expr. Is that what you had in mind? Could you recommend a way to go from hir Expr to hir Ty (currently the lint is using rustc_middle::Ty)?

Reply:
Understood! Thank you!